File: src/solutions/day18/solution.py
from collections import deque

with open('input.txt') as f:
    byte_locs = [(int(l.rstrip().split(',')[0]), int(l.rstrip().split(',')[1])) for l in f.readlines()]

grid = [['.'] * 71 for _ in range(71)]

for (x, y) in byte_locs[:1024]:
    grid[y][x] = "#"

q = deque([(0, 0, 0)])
res = float("inf")
seen = {(0, 0)}
rows, cols = len(grid), len(grid[0])

while q:
    for _ in range(len(q)):
        x, y, s = q.popleft()

        if x == rows - 1 and y == cols - 1:
            res = min(res, s)
        
        for dx, dy in [(1, 0), (-1, 0), (0, 1), (0, -1)]:
            new_x, new_y = x + dx, y + dy
            if 0 <= new_x < rows and 0 <= new_y < cols and grid[new_x][new_y] != "#" and (new_x, new_y) not in seen:
                seen.add((new_x, new_y))
                q.append((new_x, new_y, s + 1))
    
print("Solution: ", res)
    
    

        

File: src/solutions/day18/solution.ts
import {readFileSync} from 'node:fs';

const contents = readFileSync('sample.txt', 'utf-8'); 
const byte_locs = [];

for (const line of contents.split("\n")) {
    const [x, y] = line.split(",");
    byte_locs.push([Number(x), Number(y)]);
}

const rows = 7;
const cols = 7;
const grid = [];

for (let i = 0; i < rows; i++) {
    const r = [];

    for (let j = 0; j < cols; j++) {
        r.push(".");
    }

    grid.push(r);
}

for (let b = 0; b < 12; b++) {
    const [x, y] = byte_locs[b];
    grid[y][x] = "#";
}

const q = [[0, 0, 0]];
let res = 1000000;
const seen = new Set(['0,0']);

while (q.length > 0) {
    let [x, y, s] = q.shift();

    if (x === rows - 1 && y === cols - 1) {
        res = Math.min(res, s);
        continue;
    }

    for (let [dx, dy] of [[1, 0], [-1, 0], [0, 1], [0, -1]]) {
        const newX = x + dx;
        const newY = y + dy;
        
        if (0 <= newX && newX < rows && 0 <= newY && newY < cols && grid[newX][newY] !== "#" && !seen.has(`${newX},${newY}`)) {
            q.push([newX, newY, s + 1]);
            seen.add(`${newX},${newY}`);
        }
    }
}

console.log("Solution: ", res);